File: projects/iffywow/elixir/test/conformance_test.exs
Code.require_file("support/conformance.exs", __DIR__)

defmodule Ithkuil.ConformanceTest do
  @moduledoc """
  Golden-vector conformance: every row of ../../conformance/*.jsonl asserted
  in both directions (see conformance/README.md and CODEC.md).
  """
  use ExUnit.Case, async: true

  alias Ithkuil.{Codec, Coord}
  import Ithkuil.TestSupport.Conformance

  describe "codec_units.jsonl" do
    test "term ser/de and integer ranking, both directions" do
      vectors = vectors("codec_units.jsonl")
      assert vectors != []

      for row <- vectors do
        assert row["kind"] == "term"
        term = json_term(row["term"])
        bytes = hex!(row["bytes"])
        integer = String.to_integer(row["integer"])

        # serialize
        assert Codec.encode_term(term) == bytes,
               "encode_term mismatch for #{inspect(row["term"])}"

        # strict decode, no trailing bytes
        assert {:ok, ^term, <<>>} = Codec.decode_term(bytes)

        # ranking both ways
        assert Codec.rank(bytes) == integer
        assert {:ok, ^bytes} = Codec.unrank(integer)
        assert Codec.term_to_integer(term) == integer
        assert {:ok, ^term} = Codec.integer_to_term(integer)
      end
    end
  end

  describe "coordinate_to_integer.jsonl" do
    test "coordinate <-> bytes <-> integer, both directions" do
      vectors = vectors("coordinate_to_integer.jsonl")
      assert vectors != []

      for row <- vectors do
        assert {:ok, coord} = Coord.from_wire(row["coordinate"])
        bytes = hex!(row["bytes"])
        integer_string = row["integer"]

        # coordinate -> bytes -> coordinate
        assert Codec.encode_word(coord) == bytes
        assert {:ok, ^coord} = Codec.decode_word(bytes)

        # coordinate -> integer string -> coordinate
        assert {:ok, ^integer_string} = Ithkuil.to_integer_string(coord)
        assert {:ok, ^coord} = Ithkuil.from_integer_string(integer_string)

        # wire round-trip is exact (vector coordinates are canonical)
        assert Coord.to_wire(coord) == row["coordinate"]
      end
    end
  end

  describe "invalid_inputs.jsonl" do
    test "every invalid input is rejected with the expected error class" do
      vectors = vectors("invalid_inputs.jsonl")
      assert vectors != []

      for row <- vectors do
        expected = String.to_atom(row["error"])

        result =
          case row["kind"] do
            "integer_string" -> Ithkuil.from_integer_string(row["value"])
            "bytes" -> row["value"] |> hex!() |> Codec.decode_word()
            "coordinate" -> Coord.new(row["value"])
          end

        assert error_class(result) == expected,
               "expected #{expected} for #{inspect(row)}, got #{inspect(result)}"
      end
    end
  end

  describe "latin_to_coordinate.jsonl" do
    test "romanization profile core-v1 vectors, both directions" do
      vectors = vectors("latin_to_coordinate.jsonl")
      assert vectors != []

      for row <- vectors do
        case row do
          %{"error" => error} ->
            expected = String.to_atom(error)
            result = Ithkuil.from_latin(row["latin"])

            assert match?({:error, {^expected, _detail}}, result),
                   "expected #{error} for #{inspect(row["latin"])}, got #{inspect(result)}"

          %{"coordinate" => wire, "canonical" => canonical} ->
            # from_latin(latin) == coordinate
            assert {:ok, coord} = Ithkuil.from_latin(row["latin"])
            assert Coord.to_wire(coord) == wire
            assert {:ok, ^wire} = Ithkuil.to_wire(coord)

            # to_latin(coordinate) == canonical (wire input canonicalized on entry)
            assert {:ok, ^canonical} = Ithkuil.to_latin(wire)
            assert {:ok, ^canonical} = Ithkuil.to_latin(coord)

            # round-trip law: from_latin(to_latin(c)) == c
            assert {:ok, ^coord} = Ithkuil.from_latin(canonical)
        end
      end
    end
  end

  describe "CODEC.md worked example" do
    test "empty word ranks to 8606843649" do
      empty = {:ithkuil_word, 1, []}
      assert Codec.encode_word(empty) == <<0x01, 0x00, 0x01, 0x02, 0x00>>
      assert Ithkuil.to_integer(empty) == {:ok, 8_606_843_649}
      assert Ithkuil.from_integer_string("8606843649") == {:ok, empty}
    end
  end
end

File: projects/iffywow/node/test/conformance.test.js
/**
 * Golden-vector conformance tests — every row of ../../conformance/*.jsonl,
 * both directions where defined. These vectors, not shared code, keep this
 * SDK in sync with the Elixir production codec and the Python reference
 * (SDK-INTERFACE.md §5: conformance/ is the arbiter).
 */

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { Buffer } from "node:buffer";

import {
  IthkuilError,
  fromWire,
  toWire,
  toWireJson,
  fromWireJson,
  toBytes,
  fromBytes,
  toInteger,
  fromInteger,
  toIntegerString,
  fromIntegerString,
  fromLatin,
  toLatin,
} from "../src/index.js";
import { serializeTerm, deserializeTerm, termFromJson, termToJson, rank, unrank } from "../src/codec.js";

// node/test/ -> ../../conformance/
function loadVectors(name) {
  const url = new URL(`../../conformance/${name}`, import.meta.url);
  return readFileSync(url, "utf8")
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0)
    .map((line) => JSON.parse(line));
}

const hexToBytes = (hex) => new Uint8Array(Buffer.from(hex, "hex"));
const bytesToHex = (bytes) => Buffer.from(bytes).toString("hex");

function assertIthkuilError(fn, code, context) {
  let threw = false;
  try {
    fn();
  } catch (err) {
    threw = true;
    assert.ok(
      err instanceof IthkuilError,
      `${context}: expected IthkuilError, got ${err && err.constructor ? err.constructor.name : typeof err}: ${err}`,
    );
    assert.equal(err.code, code, `${context}: expected code ${code}, got ${err.code} (${err.message})`);
    assert.ok(
      err.message.startsWith(`${code}:`),
      `${context}: message must start with "${code}:", got "${err.message}"`,
    );
  }
  assert.ok(threw, `${context}: expected IthkuilError ${code}, but nothing was thrown`);
}

// ---------------------------------------------------------------------------
// codec_units.jsonl — term-level ser/rank vectors (hand-computed, normative).
// ---------------------------------------------------------------------------

for (const row of loadVectors("codec_units.jsonl")) {
  test(`codec_units: bytes ${row.bytes} (${JSON.stringify(row.term)})`, () => {
    const term = termFromJson(row.term);
    const data = hexToBytes(row.bytes);

    // serialize / deserialize
    assert.equal(bytesToHex(serializeTerm(term)), row.bytes);
    assert.deepEqual(termToJson(deserializeTerm(data)), row.term);

    // rank / unrank (BigInt end to end; decimal string at the boundary)
    assert.equal(rank(data).toString(10), row.integer);
    assert.equal(bytesToHex(unrank(BigInt(row.integer))), row.bytes);

    // JSON helper round-trip keeps the vector format itself honest
    assert.deepEqual(termToJson(termFromJson(termToJson(term))), row.term);
  });
}

// ---------------------------------------------------------------------------
// coordinate_to_integer.jsonl — word-level coordinate <-> bytes <-> integer.
// ---------------------------------------------------------------------------

for (const row of loadVectors("coordinate_to_integer.jsonl")) {
  test(`coordinate_to_integer: integer ${row.integer}`, () => {
    const data = hexToBytes(row.bytes);
    const coord = fromWire(row.coordinate);

    // encode direction
    assert.equal(bytesToHex(toBytes(coord)), row.bytes);
    assert.equal(toInteger(coord).toString(10), row.integer);
    assert.equal(toIntegerString(coord), row.integer);

    // decode direction (leading zeros preserved, strict parse)
    assert.deepEqual(fromBytes(data), coord);
    assert.deepEqual(fromInteger(BigInt(row.integer)), coord);
    assert.deepEqual(fromIntegerString(row.integer), coord);

    // the vector's coordinate is already canonical
    assert.deepEqual(toWire(coord), row.coordinate);

    // wire JSON text round-trip
    assert.equal(toWireJson(coord), JSON.stringify(row.coordinate));
    assert.deepEqual(fromWireJson(JSON.stringify(row.coordinate)), coord);
  });
}

// ---------------------------------------------------------------------------
// invalid_inputs.jsonl — inputs that MUST be rejected, by error code.
// ---------------------------------------------------------------------------

for (const [i, row] of loadVectors("invalid_inputs.jsonl").entries()) {
  test(`invalid_inputs[${i}]: ${row.kind} -> ${row.error}`, () => {
    const context = `${row.kind} ${JSON.stringify(row.value)}`;
    if (row.kind === "integer_string") {
      assertIthkuilError(() => fromIntegerString(row.value), row.error, context);
    } else if (row.kind === "bytes") {
      assertIthkuilError(() => fromBytes(hexToBytes(row.value)), row.error, context);
    } else if (row.kind === "coordinate") {
      assertIthkuilError(() => fromWire(row.value), row.error, context);
      assertIthkuilError(() => fromWireJson(JSON.stringify(row.value)), row.error, `${context} (JSON text)`);
    } else {
      assert.fail(`unknown vector kind ${JSON.stringify(row.kind)}`);
    }
  });
}

// ---------------------------------------------------------------------------
// latin_to_coordinate.jsonl — romanization profile core-v1 (ROMANIZATION.md §10).
// ---------------------------------------------------------------------------

for (const [i, row] of loadVectors("latin_to_coordinate.jsonl").entries()) {
  if (row.error !== undefined) {
    test(`latin_to_coordinate[${i}]: ${JSON.stringify(row.latin)} -> ${row.error}`, () => {
      assertIthkuilError(() => fromLatin(row.latin), row.error, `latin ${JSON.stringify(row.latin)}`);
    });
  } else {
    test(`latin_to_coordinate[${i}]: ${JSON.stringify(row.latin)}`, () => {
      const coord = fromLatin(row.latin);

      // from_latin(latin) == coordinate (wire deep-equality)
      assert.deepEqual(toWire(coord), row.coordinate);

      // to_latin(coordinate) == canonical
      assert.equal(toLatin(fromWire(row.coordinate)), row.canonical);
      assert.equal(toLatin(coord), row.canonical);

      // canonical spelling round-trips exactly
      assert.deepEqual(toWire(fromLatin(row.canonical)), row.coordinate);
      assert.equal(toLatin(fromLatin(row.canonical)), row.canonical);
    });
  }
}
